File: C#/BluffinMuffin.Protocol/AbstractCommand.cs
using System;
using System.Linq;
using System.Reflection;
using BluffinMuffin.Protocol.DataTypes.Attributes;
using BluffinMuffin.Protocol.Enums;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using Newtonsoft.Json.Linq;

namespace BluffinMuffin.Protocol
{
    /// <summary>
    /// 
    /// </summary>
    public abstract class AbstractCommand
    {
        /// <summary>
        /// Always contains '{CommandName}' to distinguish the command from others."
        /// </summary>
        [JsonProperty(Order = -100)]
        [ExampleValue("{CommandName}")]
        public string CommandName => GetType().Name;

        /// <summary>
        /// 
        /// </summary>
        [JsonIgnore]
        public abstract BluffinCommandEnum CommandType { get; }

        /// <summary>
        /// Browsing all Types inheriting "AbstractBluffinCommand", it finds the type named exactly like the "CommandName" attribute in the JSON.
        /// </summary>
        public static AbstractCommand DeserializeCommand(string data)
        {
            var cmd = JsonConvert.DeserializeObject<AbstractCommand>(data);
            Type commType = Assembly.GetExecutingAssembly().GetTypes().Single(t => t.IsClass && !t.IsAbstract && t.IsSubclassOf(typeof(AbstractCommand)) && t.Name == cmd.CommandName);
            MethodInfo method = typeof(JsonConvert).GetMethods().First(m => m.Name == "DeserializeObject" && m.IsGenericMethod).MakeGenericMethod(commType);
            return (AbstractCommand)method.Invoke(null, new object[] { data });
        }

        /// <summary>
        /// 
        /// </summary>
        /// <returns></returns>
        public string Encode()
        {
            return JsonConvert.SerializeObject(this, new StringEnumConverter());
        }
    }
}

File: Python/bluffinmuffin/protocol/interfaces/abstract_command.py
from collections import OrderedDict
import json

from ..enums.bluffin_command_enum import BluffinCommandEnum


class AbstractCommand:

    def __init__(self, command_type):
        self.command_name = self.__class__.__name__
        self.command_type = command_type

    def __str__(self):
        return '[{0}] {1}'.format(BluffinCommandEnum.to_char(self.command_type), self.command_name)

    def _encode_specific(self, d):
        return None

    def _encode_specific_end(self, d):
        return None

    def encode(self):
        d = OrderedDict()
        d['CommandName'] = self.command_name
        self._encode_specific(d)
        self._encode_specific_end(d)
        #return json.dumps(d, sort_keys=False, indent=4, separators=(',', ': '))
        return json.dumps(d, sort_keys=False)
